BDArmory/Guidances/MissileGuidance.cs

```diff
@@ -147,6 +147,184 @@ public static Vector3 GetAirToGroundTarget(Vector3 targetPosition, Vector3 targe
+                bulletRelativePosition = targetPosition - firePosition;
+                float timeToCPA = timeToImpact; // Rough initial estimate.
+                targetPredictedPosition = AIUtils.PredictPosition(targetPosition, targetVelocity, targetAcceleration, timeToCPA);
+
+                // Velocity Compensation Logic
+                float compMult = Mathf.Clamp(0.5f * (targetDistance - termDist) / termDist, 0f, 1f);
+                Vector3 velDirectionHor = (velDirection - upDirection * Vector3.Dot(velDirection, upDirection)).normalized;
+                targetHorVel = targetVelocity - upDirection * Vector3.Dot(targetVelocity, upDirection); // Get target horizontal velocity (relative to missile frame)
+                float targetAlVelMag = Vector3.Dot(targetHorVel, velDirectionHor); // Get magnitude of velocity aligned with the missile velocity vector (in the horizontal axis)
+                targetAlVelMag *= Mathf.Sign(velComp) * compMult;
+                targetAlVelMag = 0.5f * (targetAlVelMag + Mathf.Abs(targetAlVelMag)); // Set -ve velocity (I.E. towards the missile) to 0 if velComp is +ve, otherwise for -ve
+
+                //targetCompVel = targetVelocity + velComp * targetHorVel.magnitude* targetHorVel.normalized; // Old velComp logic
+                targetCompVel = targetVelocity + velComp * targetAlVelMag * velDirectionHor; // New velComp logic
+
+                var count = 0;
+                do
+                {
+                    lastVelDirection = velDirection;
+                    currVel = currSpeed * velDirection;
+                    firePosition = missileVessel.transform.position + (currSpeed * velDirection) * Time.fixedDeltaTime; // Bullets are initially placed up to 1 frame ahead (iTime).
+                    bulletAcceleration = FlightGlobals.getGeeForceAtPosition((firePosition + targetPredictedPosition) / 2f); // Drag is ignored.
+                    bulletRelativePosition = targetPosition - firePosition + compMult * altComp * upDirection; // Compensate for altitude
+                    bulletRelativeVelocity = targetVelocity - currVel;
+                    bulletRelativeAcceleration = targetAcceleration - bulletAcceleration;
+                    timeToCPA = AIUtils.TimeToCPA(bulletRelativePosition, bulletRelativeVelocity, bulletRelativeAcceleration, timeToImpact*3f);
+                    targetPredictedPosition = AIUtils.PredictPosition(targetPosition, targetCompVel, targetAcceleration, timeToCPA);
+                    bulletDropOffset = -0.5f * bulletAcceleration * timeToCPA * timeToCPA;
+                    ballisticTarget = targetPredictedPosition + bulletDropOffset;
+                    velDirection = (ballisticTarget - missileVessel.transform.position).normalized;
+                } while (++count < 10 && Vector3.Angle(lastVelDirection, velDirection) > 1f); // 1° margin of error is sufficient to prevent premature firing (usually)
```

Comment:
You probably don't need as many as 10 iterations here, since you don't need to be as precise as for aiming a gun due to being able to steer during final guidance.

It's likely that you don't even need to use an iteration, since the missile is likely heading in roughly the correct direction already for the `timeToCPA` to be fairly accurate and if it's adjusting its `velDirection` for the `ballisticTarget` over time, then it's effectively doing the iteration over multiple frames (it would only fail to converge if the missile is heading away from the target such that the `timeToCPA` would be 0).

Reply:
Tested with 5 iterations and the guidance is slightly worse, the effectiveness of this guidance method is very dependent on this pre-terminal midcourse guidance so having this be accurate is fairly important, particularly as the trigger for the pre-terminal midcourse guidance is also based on the result from this.

Reply:
OK